<launch>
  <!-- 
  bringup_sim.launch 
  
  This launch file starts the flyer software for the simulated flyer.
  
  Usage:
  First, start a roscore and an rxconsole on the ground station.
  Then, on the ground station:
  $ roslaunch flyer_common bringup_sim.launch
  
  By default plots of various key quantities are launched on the
  ground station, as well as a runtime_monitor to check on diagnostics.
  
  The 'hover' control mode is brought to standby automatically.
  -->

  <!-- To debug a node (running locally) in GDB in an xterm, add the attribute: -->
  <!-- launch-prefix="xterm -rv -e gdb -ex run -args" -->

  <!-- ARGUMENTS -->
  <arg name="machine" default="local"/>
  <arg name="flyer_name" default="simflyer1"/>
  <arg name="model" default="simple_nonlinear"/>
  <arg name="external_frame_heading" default="90.0"/>
  <arg name="viz" default="true"/>
  <arg name="rviz_vcg_file" default="$(find starmac_viz)/flyer_viz.vcg"/>
  <arg name="sim_publish_odometry" default="false"/>
  <arg name="sim_odom_topic" default="odom"/>
  <arg name="sim_publish_state" default="true"/>
  <arg name="sim_update_rate" default="120.0"/>
  <!-- ground station: -->
  <arg name="show_plots" default="true"/>
  <arg name="monitoring" default="true"/>
  <arg name="odom_topic" default="estimator/output"/>
  <arg name="viz_odom_topic" default="$(arg odom_topic)"/>
  <arg name="dummy_joystick" default="false"/>
  <arg name="soft_joystick" default="true"/>
  <arg name="start_estimator" default="true"/>
  <arg name="standby_modes" default="hover"/>
  <arg name="nodelets_standalone" default="false"/>
  <!-- logging: -->
  <arg name="record_all" default="false"/>
  <arg name="record_state_and_inputs" default="false"/>
  <arg name="tag" default=""/>
  
  <!-- "VARIABLES" -->
  <arg name="flyer_machine" value="local"/>
  <arg name="nodelet_manager" value="manager"/>
  <arg name="run_nodelet_manager_in_gdb" default="false"/>

  <!-- MACHINE DEFINITIONS -->
  <group>
    <include file="$(find flyer_common)/launch/machines/local.xml"/>
    <include file="$(find flyer_common)/launch/machines/sim_flyers.xml"/>
  </group>
  
  <group ns="$(arg flyer_name)">
    <!--  machine name="local" address="localhost"/>-->

    <remap from="odom" to="$(arg odom_topic)"/>
    
    <!-- FRAMES -->
    <include file="$(find flyer_common)/launch/systems/frames.xml">
    </include>
    
    <!-- TUI -->
    <include file="$(find flyer_common)/launch/systems/tui.xml"/>
    
    <!-- JOYSTICK -->
    <include file="$(find flyer_common)/launch/systems/joystick.xml">
      <arg name="dummy_joystick" value="$(arg dummy_joystick)"/>
      <arg name="soft_joystick" value="$(arg soft_joystick)"/>
    </include>
        
    <!-- NODELET MANAGER -->
    <include unless="$(arg nodelets_standalone)"
        file="$(find flyer_common)/launch/systems/nodelet_manager.xml">
        <arg name="name" value="$(arg nodelet_manager)"/>
        <arg name="num_worker_threads" value="40"/>
        <arg name="machine" value="$(arg flyer_machine)" />
        <arg name="run_in_gdb" value="$(arg run_nodelet_manager_in_gdb)"/>
    </include>
        
    <!-- CONTROL -->
    <include file="$(find flyer_common)/launch/systems/control.xml">
      <arg name="machine" value="$(arg flyer_machine)"/>
      <arg name="teleop_use_udp" value="false"/>
      <arg name="nodelet_manager" value="manager"/>
      <arg name="teleop_joy_topic" value="joy"/>
      <arg name="nodelets_standalone" default="$(arg nodelets_standalone)"/>
    </include>
    
    <!--  BRING HOVER MODE TO STANDBY -->
    <!-- node pkg="rosh" type="rosh" name="standby_rosh" args="flyer_common/modes_to_standby.py" machine="$(arg flyer_machine)">
      <param name="modes" value="$(arg standby_modes)"/>
    </node -->
     <node pkg="flyer_common" type="modes_to_standby_nonrosh.py" name="modes_to_standby" machine="$(arg flyer_machine)">
      <param name="modes" value="$(arg standby_modes)"/>
    </node>
    
    <!-- ESTIMATOR -->
    <arg if="$(arg nodelets_standalone)" name="nodelet_arg1" value="standalone" />
    <arg if="$(arg nodelets_standalone)" name="nodelet_arg3" value="" />
    <arg unless="$(arg nodelets_standalone)" name="nodelet_arg1" value="load" />
    <arg unless="$(arg nodelets_standalone)" name="nodelet_arg3" value="$(arg nodelet_manager) --no-bond" />
    
    <node name="estimator" pkg="nodelet" type="nodelet"
          args="$(arg nodelet_arg1) flyer_est/Estimator $(arg nodelet_arg3)" machine="$(arg flyer_machine)">
      <rosparam>
        freq: 120
        # following filter parameters tuned for 120Hz Vicon rate
        xy_vel_filt_a: 0.2
        xy_vel_filt_b: 0.8
        z_vel_filt_a: 0.2
        z_vel_filt_b: 0.8
        ang_vel_filt_a: 0.2
        ang_vel_filt_b: 0.8
      </rosparam>
      <param name="transform_topic" value="state"/>
    </node>
    
    <!-- STATIC TRANSFORMS -->
    <node pkg="flyer_common" type="static_transforms.py" name="static_transforms" machine="$(arg flyer_machine)">
      <param name="mode" value="both" type="str"/>
      <param name="flyer" value="$(arg flyer_name)" type="str"/>
      <param name="period" value="2400"/>
    </node>
    
    <!-- RELAYS -->
    <!-- include file="$(find flyer_common)/launch/systems/downlink_relays.xml"/ -->

    <!-- MONITORING -->
    <include if="$(arg monitoring)" file="$(find flyer_common)/launch/systems/monitoring.xml">
      <arg name="diag_topic_prefix" value="/"/>
    </include>

    <!-- PLOTTING -->
    <include if="$(arg show_plots)" file="$(find flyer_common)/launch/systems/plotting.xml">
      <arg name="odom_topic" value="$(arg odom_topic)"/>
    </include>

    <!-- RECORDING -->
    <group>
      <include if="$(arg record_all)" file="$(find flyer_common)/launch/logging/all.xml">
        <arg name="machine" value="$(arg flyer_machine)"/>
        <arg name="tag" value="$(arg tag)"/>
      </include>

      <include if="$(arg record_state_and_inputs)" file="$(find flyer_common)/launch/logging/state_and_inputs.xml">
        <arg name="machine" value="$(arg flyer_machine)"/>
        <arg name="tag" value="$(arg tag)"/>
      </include>
    </group>
    
    <!-- SIMULATION -->
    <include file="$(find simulated_quadrotor)/launch/sim.xml">
      <arg name="model" value="$(arg model)"/>
      <arg name="rate" value="$(arg sim_update_rate)"/>
      <arg name="publish_odometry" value="$(arg sim_publish_odometry)"/>
      <arg name="sim_odom_topic" value="$(arg sim_odom_topic)"/>
      <arg name="publish_state" value="$(arg sim_publish_state)"/>
    </include>

    <!-- VISUALIZATION -->
    <include if="$(arg viz)" file="$(find flyer_common)/launch/systems/visualization.xml">
      <arg name="vcg_file" value="$(arg rviz_vcg_file)"/>
      <arg name="odom_topic" value="$(arg viz_odom_topic)"/>
    </include>
    
    <param name="control_mode_attitude/external_command_frame" value="true"/>    
  
    <!-- PARAMETER OVERRIDES -->
    <param name="control_mode_attitude/external_command_frame" value="true"/>
    <param name="control_mode_attitude/external_frame_heading" type="double" value="$(arg external_frame_heading)"/>
    <param name="control_mode_hover/external_command_frame" value="true"/>
    <param name="control_mode_hover/external_frame_heading" type="double" value="$(arg external_frame_heading)"/>
  
  </group>

</launch>